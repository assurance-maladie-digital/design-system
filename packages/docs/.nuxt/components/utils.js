// nuxt/nuxt.js#8607
export function wrapFunctional(options) {
  if (!options || !options.functional) {
    return options
  }

  const propKeys = Array.isArray(options.props) ? options.props : Object.keys(options.props || {})

  return {
    render(h) {
      const attrs = {}
      const props = {}

      for (const key in this.$attrs) {
        if (propKeys.includes(key)) {
          props[key] = this.$attrs[key]
        } else {
          attrs[key] = this.$attrs[key]
        }
      }

      return h(options, {
        on: this.$listeners,
        attrs,
        props,
        scopedSlots: this.$scopedSlots,
      }, this.$slots.default)
    }
  }
}
