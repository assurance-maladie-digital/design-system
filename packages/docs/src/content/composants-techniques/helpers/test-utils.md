---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
