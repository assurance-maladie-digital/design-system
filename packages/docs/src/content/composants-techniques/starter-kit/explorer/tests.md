Le dossier `tests` contient les tests d’interface et les tests unitaires globaux de l’application.

<doc-alert type="warning">
Il n’est pas recommandé de renommer ce dossier.
</doc-alert>
