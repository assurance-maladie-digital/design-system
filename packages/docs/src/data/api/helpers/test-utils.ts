import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
