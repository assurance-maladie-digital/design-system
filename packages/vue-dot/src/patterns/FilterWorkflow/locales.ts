export const locales = {
	addFilterLabel: 'filtres',
	nbFilters:  (nbFilters: number): string => `${nbFilters} filtre${nbFilters > 1 ? 's' : ''}`,
	applyFilterLabel: 'Appliquer le filtre'
};
