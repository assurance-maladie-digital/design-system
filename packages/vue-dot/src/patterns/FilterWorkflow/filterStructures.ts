import { parseDate } from '../../helpers/parseDate';
import { PeriodValue, FieldValue } from '@cnamts/form-builder/src/components/FormField/types.d';
import { FilterStructures, ColumnValue } from './types';
import { ChoiceValue, FieldItemValue } from '@cnamts/form-builder/src/components/FormField/types';
import { FormValues } from '@cnamts/form-builder/src/functions/getFormValues/types';
import dayjs from 'dayjs';

export type FilterValueText<T = FormValues> = (formValues: T) => string | null;

export const filterStructures: FilterStructures = {
	text: {
		fields: {
			select: {
				type: 'select',
				fieldOptions: {
					type: 'choiceAutocomplete',
					outlined: true,
					label: 'Rechercher',
					menuProps: {
						origin: 'bottom',
						maxHeight: 185,
						disableKeys: true
					},
					class: 'pb-16 mb-16'
				},
				value: null,
				multiple: true
			}
		},
		valueToString: (formValues: FormValues): string | null => {
			const fieldSelect = formValues.select as ChoiceValue;

			const selectedItems = fieldSelect.value as FieldItemValue[];

			if (!selectedItems || !selectedItems.length) {
				return null;
			}

			let text = '';
			const nbSelectedItems = selectedItems.length;

			if (nbSelectedItems > 1) {
				text = `${nbSelectedItems} éléments sélectionnés`;
			} else {
				text = String(selectedItems[0]);
			}

			return text;
		},
		filter: (value: ColumnValue, formValues: FormValues): boolean => {
			const fieldSelect = formValues.select as ChoiceValue;

			const selectedItems = fieldSelect?.value as FieldItemValue[];

			if (selectedItems && selectedItems.length) {
				return selectedItems.includes(value);
			}

			return true;
		}
	},
	date: {
		fields: {
			date: {
				type: 'period',
				fieldOptions: {
					from: {
						outlined: true,
						textField: {
							label: 'A partir du'
						}
					},
					to: {
						outlined: true,
						textField: {
							label: 'Jusqu\'au'
						}
					}
				}
			}
		},
		valueToString: (formValues: FormValues): string | null => {
			const filterValue = formValues.date as PeriodValue;

			const fromValue = filterValue?.from;
			const toValue = filterValue?.to;

			let text = '';

			// Check if there is a filter to apply
			if (!fromValue && !toValue) {
				return null;
			}

			if (fromValue) {

				text = `A partir du ${dayjs(fromValue).format('DD/MM/YYYY')}`;
			}

			if (fromValue && toValue) {
				text += ', ';
			}

			if (toValue) {
				text += `Jusqu'au ${dayjs(toValue).format('DD/MM/YYYY')}`;
			}

			return text;
		},
		filter: (value: ColumnValue, formValues: FormValues): boolean => {
			const filterValue = formValues.date as PeriodValue;

			const fromValue = filterValue?.from;
			const toValue = filterValue?.to;

			// Check if there is a filter to apply
			if (!fromValue && !toValue) {
				return true;
			}

			// Filter column with value equal to null or not string value
			if (!value || typeof value !== 'string') {
				return false;
			}

			// Check from value
			if (
				fromValue && parseDate(value as string).toDate() < dayjs(fromValue).toDate()
			) {
				return false;
			}

			// Check toValue
			if (
				toValue && parseDate(value as string).toDate() > dayjs(toValue).toDate()
			) {
				return false;
			}

			return true;
		}
	},
	number: {
		fields: {
			numberMin: {
				type: 'number',
				fieldOptions: {
					outlined: true,
					mask: '###########',
					label: 'Supérieur ou égal'
				}
			},
			numberMax: {
				type: 'number',
				fieldOptions: {
					outlined: true,
					mask: '###########',
					label: 'Inférieur ou égal'
				}
			}
		},
		valueToString: (formValues: FormValues): string | null => {
			const numberMin = formValues.numberMin as FieldValue;
			const numberMax = formValues.numberMax as FieldValue;
			let text = '';

			// Check if there is a filter to apply
			if (!numberMax && !numberMin) {
				return null;
			}

			if (numberMin) {
				text = `supérieur à ${numberMin}`;
			}

			if (numberMax && numberMin) {
				text += ', ';
			}

			if (numberMax) {
				text += `inférieur à ${numberMax}`;
			}

			return text;
		},
		filter: (value: ColumnValue, formValues: FormValues): boolean => {
			const numberMin = formValues.numberMin as string;
			const numberMax = formValues.numberMax as string;

			// Check if there is a filter to apply
			if (!numberMax && !numberMin) {
				return true;
			}

			// Filter column with value equal
			if (!value) {
				return false;
			}

			// Check numberMin
			if (numberMin && Number.parseInt(value as string, 10) < Number.parseInt(numberMin, 10)) {
				return false;
			}

			// Check numberMax
			if (numberMax && Number.parseInt(value as string, 10) > Number.parseInt(numberMax, 10)) {
				return false;
			}

			return true;
		}
	}
};
