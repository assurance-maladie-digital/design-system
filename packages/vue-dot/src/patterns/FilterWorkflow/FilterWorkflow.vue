<template>
	<div>
		<VMenu
			v-if="activeFilters.length"
			offset-y
			:close-on-content-click="false"
		>
			<template #activator="{ on }">
				<VChip
					class="mx-1"
					color="accent"
					close
					v-on="on"
					@click:close="deleteAllFilters"
				>
					<span>
						{{ locales.nbFilters(activeFilters.length) }}
					</span>
				</VChip>
			</template>

			<VList
				width="400"
				class="py-0"
			>
				<VListItem
					v-for="(row, index) in activeFilters"
					:key="index"
					two-line
				>
					<VListItemContent class="mr-2">
						<VListItemTitle v-text="row.label" />

						<VListItemSubtitle
							class="body-2"
							v-text="row.text"
						/>
					</VListItemContent>

					<VListItemAction>
						<VBtn
							icon
							@click="openFilterDialog(row.filterName)"
						>
							<VIcon>
								{{ editIcon }}
							</VIcon>
						</VBtn>
					</VListItemAction>

					<VListItemAction class="ml-0">
						<VBtn
							icon
							@click="deleteFilter(index)"
						>
							<VIcon>
								{{ removeIcon }}
							</VIcon>
						</VBtn>
					</VListItemAction>
				</VListItem>
			</VList>
		</VMenu>

		<VMenu offset-y>
			<template #activator="{ on }">
				<VBtn
					text
					class="mr-8"
					v-on="on"
				>
					<VIcon>{{ filterIcon }}</VIcon>
					{{ locales.addFilterLabel }}
				</VBtn>
			</template>

			<VList
				width="200"
				class="py-0"
			>
				<VListItem
					v-for="(row, index) in filters"
					:key="index"
					@click="openFilterDialog(index)"
				>
					<VListItemTitle v-text="row.label" />
				</VListItem>
			</VList>
		</VMenu>

		<VDialog
			v-model="filterDialog"
			max-width="360"
		>
			<VCard v-if="filterTypeEdit">
				<VCardTitle v-if="filterEditLabel">
					{{ filterEditLabel }}
				</VCardTitle>

				<VCardText class="pt-4">
					<VForm ref="filterForm">
						<FormFieldList
							:fields="filterTypeEdit.fields"
							@change="fieldsUpdated($event)"
						/>
					</VForm>

					<VBtn
						class="mt-8"
						block
						color="accent"
						@click="applyFilter"
					>
						{{ locales.applyFilterLabel }}
					</VBtn>
				</VCardText>
			</VCard>
		</VDialog>
	</div>
</template>

<script lang="ts">
	import Component, { mixins } from 'vue-class-component';

	import { mdiFilterVariant, mdiPencil, mdiClose } from '@mdi/js';

	import { Refs } from '@cnamts/vue-dot/src/types';

	import { locales } from './locales';

	import { FilterWorkflowCore } from './mixins/filterWorkflowCore';
	import FormFieldList from '@cnamts/form-builder/src/components/FormFieldList/FormFieldList.vue';

	const MixinsDeclaration = mixins(FilterWorkflowCore);

	@Component<FilterWorkflow>({
		components: {
			FormFieldList
		}
	})
	export default class FilterWorkflow extends MixinsDeclaration {
		$refs!: Refs<{
			filterForm: HTMLFormElement;
		}>;

		filterIcon = mdiFilterVariant;
		editIcon = mdiPencil;
		removeIcon = mdiClose;

		locales = locales;

		/** Apply the editing filter */
		submitForm(): void {
			if (!this.filterEditName || !this.filterTypeEdit) {
				return;
			}

			if (!this.$refs.filterForm.validate()) {
				return;
			}

			this.applyFilter();
		}
	}
</script>
