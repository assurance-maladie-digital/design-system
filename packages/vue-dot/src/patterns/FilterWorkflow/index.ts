import FilterWorkflow from './FilterWorkflow.vue';

export default FilterWorkflow;
