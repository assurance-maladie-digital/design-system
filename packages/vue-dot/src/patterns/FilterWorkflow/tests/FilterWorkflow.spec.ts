import { dataRows, dataFilters } from './data/dataFilters';
import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';
import { html } from '@/tests/utils/html';

import FilterWorkflow from '..';

let wrapper: Wrapper<Vue>;

// Tests
describe('FilterWorkflow', () => {
	it('renders correctly', () => {
		// Mount component
		wrapper = mountComponent(FilterWorkflow, {
			propsData: {
				rows: dataRows,
				filters: dataFilters
			}
		});

		expect(html(wrapper)).toMatchSnapshot();
	});

	it('show multiple actives filters', () => {
		// Mount component
		wrapper = mountComponent(FilterWorkflow, {
			propsData: {
				rows: dataRows,
				filters: dataFilters
			},
			mocks: {
				activeFilters: [
					{
						filterName: 'date',
						label: 'date',
						text: 'text',
						type: 'date',
						values: {
							from: '2020/12/01'
						}
					},
					{
						filterName: 'nbDownload',
						label: 'Nombre de téléchargement',
						text: 'minimum 1',
						type: 'number',
						values: {
							min: '1'
						}
					}
				]
			}
		}, true);

		// TODO: set multiple filter active

		expect(html(wrapper)).toMatchSnapshot();
	});
});
