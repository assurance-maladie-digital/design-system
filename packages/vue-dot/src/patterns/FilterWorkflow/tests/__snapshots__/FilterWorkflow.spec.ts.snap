// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterWorkflow renders correctly 1`] = `
<div>
  <!---->
  <vmenu-stub opendelay="0" closedelay="0" contentclass="" maxwidth="auto" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" openonclick="true" closeonclick="true" closeoncontentclick="true" maxheight="auto" offsety="true" origin="top left" transition="v-menu-transition">
    <vlist-stub width="200" tag="div" class="py-0">
      <vlistitem-stub activeclass="" tag="div">
        <vlistitemtitle-stub>Fichier</vlistitemtitle-stub>
      </vlistitem-stub>
      <vlistitem-stub activeclass="" tag="div">
        <vlistitemtitle-stub>Nombre de téléchargement</vlistitemtitle-stub>
      </vlistitem-stub>
      <vlistitem-stub activeclass="" tag="div">
        <vlistitemtitle-stub>Date</vlistitemtitle-stub>
      </vlistitem-stub>
    </vlist-stub>
  </vmenu-stub>
  <vdialog-stub opendelay="0" closedelay="0" contentclass="" maxwidth="360" origin="center center" retainfocus="true" transition="dialog-transition" width="auto">
    <!---->
  </vdialog-stub>
</div>
`;

exports[`FilterWorkflow show multiple actives filters 1`] = `
<div>
  <!---->
  <div class="v-menu"><button type="button" class="mr-8 v-btn v-btn--flat v-btn--text theme--light v-size--default"><span class="v-btn__content"><span aria-hidden="true" class="v-icon notranslate theme--light"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg"><path d="M6,13H18V11H6M3,6V8H21V6M10,18H14V16H10V18Z"></path></svg></span>
      filtres
      </span></button>
    <!---->
  </div>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</div>
`;
